core/src/main/java/org/apache/iceberg/io/ContentCache.java

```diff
@@ -0,0 +1,310 @@
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.PropertyUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ContentCache {
+  private static final Logger LOG = LoggerFactory.getLogger(ContentCache.class);
+  private static final int BUFFER_CHUNK_SIZE = 4 * 1024 * 1024; // 4MB
+
+  public static ContentCache createCache(Map<String, String> properties) {
+    boolean enabled =
+        PropertyUtil.propertyAsBoolean(
+            properties,
+            CatalogProperties.IO_CACHE_ENABLED,
+            CatalogProperties.IO_CACHE_ENABLED_DEFAULT);
+    long durationMs =
+        PropertyUtil.propertyAsLong(
+            properties,
+            CatalogProperties.IO_CACHE_EXPIRATION_INTERVAL_MS,
+            CatalogProperties.IO_CACHE_EXPIRATION_INTERVAL_MS_DEFAULT);
+    long totalBytes =
+        PropertyUtil.propertyAsLong(
+            properties,
+            CatalogProperties.IO_CACHE_MAX_TOTAL_BYTES,
+            CatalogProperties.IO_CACHE_MAX_TOTAL_BYTES_DEFAULT);
+    long contentLength =
+        PropertyUtil.propertyAsLong(
+            properties,
+            CatalogProperties.IO_CACHE_MAX_CONTENT_LENGTH,
+            CatalogProperties.IO_CACHE_MAX_CONTENT_LENGTH_DEFAULT);
+
+    if (!enabled) {
+      LOG.debug("No ContentCache created. {} is false.", CatalogProperties.IO_CACHE_ENABLED);
+    } else if (durationMs < 0) {
+      LOG.debug(
+          "No ContentCache created. {} ({}) is less than 0.",
+          CatalogProperties.IO_CACHE_EXPIRATION_INTERVAL_MS,
+          durationMs);
+    } else if (totalBytes <= 0) {
+      LOG.debug(
+          "No ContentCache created. {} ({}) is equal or less than 0.",
+          CatalogProperties.IO_CACHE_MAX_TOTAL_BYTES,
+          totalBytes);
+    } else if (contentLength <= 0) {
+      LOG.debug(
+          "No ContentCache created. {} ({}) is equal or less than 0.",
+          CatalogProperties.IO_CACHE_MAX_CONTENT_LENGTH,
+          contentLength);
+    } else {
+      return new ContentCache(durationMs, totalBytes, contentLength);
+    }
+
+    return null;
+  }
+
+  private final long expireAfterAccessMs;
+  private final long maxTotalBytes;
+  private final long maxContentLength;
+  private final Cache<String, CacheEntry> cache;
+
+  public ContentCache(long expireAfterAccessMs, long maxTotalBytes, long maxContentLength) {
+    this.expireAfterAccessMs = expireAfterAccessMs;
+    this.maxTotalBytes = maxTotalBytes;
+    this.maxContentLength = maxContentLength;
+
+    Caffeine<Object, Object> builder = Caffeine.newBuilder();
+    if (expireAfterAccessMs > 0) {
+      builder = builder.expireAfterAccess(Duration.ofMillis(expireAfterAccessMs));
+    }
+
+    this.cache =
+        builder
+            .maximumWeight(maxTotalBytes)
```

Comment:
should we include weak keys/values here as well?

Reply:
Added `softValues()` to this builder.

The key still use strong reference to allow key matching by String `equals()` rather than object match (`==`).